### PR TITLE
Build SFED historical baseline

### PR DESCRIPTION
**Pull request by zackarno**
built proof of concept code/funcs to create daily COGS w/ 2 bands: SFED & SFED_BASELINE. Code is contained & sourced from: `exploratory/02_wrangle_floodscan.py`

### CODE REVIEW

**Comment by hannahker** [comment]
Some initial notes on extra things I needed to get the `02_wrangle_floodscan.py` notebook to run in full: 
- `pipreqs` might not work fully here :( I also had to install `dask`, `scipy`,  `netcdf4`, and got an error that `azure_storage` is deprecated, so `azure-storage-blob` is the appropriate replacement. 
- I'd suggest using a (gitignored)`.env` file for managing environment variables, alongside the `python-dotenv` package. The main `README` can then document what the contents of that file should be. 
- We should add `pip install -e.` to the setup requirements in the main readme's (because I keep forgetting...)

**Comment by zackarno** [comment]
> * `pipreqs` might not work fully here :( I also had to install `dask`, `scipy`,  `netcdf4`, and got an error that `azure_storage` is deprecated, so `azure-storage-blob` is the appropriate replacement.

Good call - definitely needed to add this by hand. And i believe a user would have to add this to a lot of our repos. Oddly `netcdf4` doesn't show up with `pip list` even though i pip installed it to the env and and the code runs on here. Would be nice if `pip-chill` worked and we could standardize around it but i can't get it to work on this python distribution (3.12.4) 

> * I'd suggest using a (gitignored)`.env` file for managing environment variables, alongside the `python-dotenv` package. The main `README` can then document what the contents of that file should be.

yup, I had an ignored `.env` file - but added info to readme

> * We should add `pip install -e.` to the setup requirements in the main readme's (because I keep forgetting...)

yes good idea - added

**Comment by hannahker** [comment]
> yup, I had an ignored .env file - but added info to readme

Ok nice! Maybe I just can't find it in the code, but you'll also need to call `load_dotenv()` (after `from dotenv import load_dotenv`) for it to work.

**Comment by zackarno** [comment]
> > yup, I had an ignored .env file - but added info to readme
> 
> Ok nice! Maybe I just can't find it in the code, but you'll also need to call `load_dotenv()` (after `from dotenv import load_dotenv`) for it to work.

good call. Yeah thanks - I had the .env file, but lost the imports on last push - fixed!

File: README.md [hunk 1]
```diff
@@ -13,6 +11,33 @@ TBD
...
+Create a virtual env & activate it. Then install the the requirements with
+
+```shells
+pip install -r requirements.txt
+```
+
+Next install module code in src using the command:
+
+```shell
+pip install -e.
```
Comment (t-downing) [inline README.md]: `pip install -e .` is just a bit clearer (with the space)

File: src/datasources/floodscan.py [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+    cogs_list = [
+        x.name
+        for x in container_client.list_blobs(name_starts_with=prefix)
+        if (gen_utils.extract_date(x.name).date() >= start_date)
+        & (gen_utils.extract_date(x.name).date() <= end_date)  # noqa
+    ]
+
+    das = []
+    for cog in cogs_list:
+        da_in = process_floodscan_cog(
```
Comment (t-downing) [inline src/datasources/floodscan.py:53]: While working off the old COGs in `global/raster/cogs/aer...` you will probably have to check that you're reading the right band. I found that the order was not consistent (i.e. usually it was `("SFED", "MFED")`, but not always, so had to write [this](https://github.com/OCHA-DAP/ds-floodexposure-monitoring/blob/7e94941c08101d0e95cd23e505aac3bacaed4edc/src/datasources/floodscan.py#L58-L67) to handle it). Otherwise sometimes you will accidentally grab the `MFED`, but it won't be obvious because the band names don't automatically get read in, just the order of the bands.

File: exploration/01_baseline_calc_10yr_doy.py [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+In the official pipeline we can link it to the historical tif record on the
+blob.
+
+Steps:
+
+1. Smooth historical data w/ 10 day rolling mean (centered)
+2. filter data to last 10 years of data (exlcusive of current year)
+3. Calculate the mean SFED value per day of year from smoothed raster stack.
+"""
+
```
Comment (t-downing) [inline exploration/01_baseline_calc_10yr_doy.py:15]: Just in case this is useful- if you add a cell with
```
%load_ext autoreload
%autoreload 2
```
then you can make edits to your imported local functions (in `src`) and they will work right away in the notebook, instead of having to restart the kernel every time you make a change in the `src` file. Works nicely with the whole `pip install -e .` method.

File: exploration/02_wrangle_floodscan.py [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+# will be in blob. Only doing a couple days just as proof of concept.
+end_date = gen_utils.date_to_run()
+start_date = end_date - datetime.timedelta(days=3)
+da_current = floodscan.load_floodscan_cogs(
+    start_date=start_date, end_date=end_date
+)
+ds_current = da_current.to_dataset(dim="band")
+ds_current_sfed = (
+    ds_current.drop_vars(1)
+    .rename_vars({2: "SFED"})
```
Comment (t-downing) [inline exploration/02_wrangle_floodscan.py]: Again, just the thing about making sure it's the right band.

File: exploration/02_wrangle_floodscan.py [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+# merge historical and current data based on date/doy.
+for date_temp in ds_current_sfed["date"].values:
+    c_sfed_temp = ds_current_sfed.sel(date=date_temp)
+    dt_temp = pd.to_datetime(date_temp)
+    doy_temp = dt_temp.dayofyear
+    h_sfed_temp = ds_historical_baseline.sel(dayofyear=doy_temp)
+
+    # floating point issue causes lat/lon values from nc to be recognized as
+    # different than the COGS in blob even though they are the 'same'
+    h_sfed_temp_interp = h_sfed_temp.interp_like(c_sfed_temp, method="nearest")
```
Comment (t-downing) [inline exploration/02_wrangle_floodscan.py:86]: Nice 😎

File: requirements.txt
Comment (t-downing) [inline requirements.txt:1]: Could probably take out `pre-commit` since it's not actually needed for running things. May want to add `netCDF` since I had to manually install- not a big deal though.

File: src/datasources/floodscan.py [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+            cog_name=cog, container_name="global", mode="dev"
+        )
+        das.append(da_in)
+    return xr.combine_by_coords(das, combine_attrs="drop")
```
Comment (t-downing) [inline src/datasources/floodscan.py:59]: If just stacking by a single dimension at once, can use `xr.concat()` instead (and then you don't need to do `da_in = da_in.expand_dims()` beforehand). But perhaps this is better anyways because I think it looks for duplicate coords.

File: src/utils/cloud_utils.py [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+    if mode == "local":
+        df.to_parquet(fname, engine="pyarrow", index=False)
+    else:
+        # Convert the DataFrame to a Parquet file in memory
+        parquet_buffer = BytesIO()
+        df.to_parquet(parquet_buffer, engine="pyarrow", index=False)
+        parquet_buffer.seek(0)  # Rewind the buffer
+        container_client = get_container_client(mode, "tabular")
+        data = parquet_buffer.getvalue()
+        container_client.upload_blob(name=fname, data=data, overwrite=True)
```
Comment (t-downing) [inline src/utils/cloud_utils.py:72]: Oh interesting is it better to do this way with a buffer? I had just been doing straight up `blob_client.upload_blob(df.to_parquet())` but definitely open to improvements.

File: exploration/01_baseline_calc_10yr_doy.py [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+days_to_plot = [1, 10, 200]
+
+fig, axes = plt.subplots(nrows=1, ncols=3, figsize=(15, 5))
+
+for ax, day in zip(axes, days_to_plot):
+    da_doy_mean.sel(dayofyear=day).plot(ax=ax)
+    ax.set_title(f"Day {day}")
+
+plt.tight_layout()
+plt.show()
```
Comment (t-downing) [inline exploration/01_baseline_calc_10yr_doy.py:57]: This didn't work for me when running in the notebook, had to just run `fig` instead to see the figure, not a big deal though.

File: exploration/01_baseline_calc_10yr_doy.py [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+
+# Create a sequence of the last 10 years, exclusive of the current year
+last_10_years = list(range(current_year - 10, current_year))
+
+# open historical netcdf file from gdrive just bc so much faster than loading
+# from the blob
+da_fs = floodscan.open_historical_floodscan()
+
+# rolling window to smooth out historical data
+da_fs_smooth = da_fs.rolling(time=10, center=True).mean()
```
Comment (t-downing) [inline exploration/01_baseline_calc_10yr_doy.py:36]: What do we think about 10 days as the window? I mean this seems about right to me (I've often used 7, partially just because having `center=True` felt nicer on an odd number). I did note that Floodscan already applies some sort of smoothing to their values though. 

I'm generally good with 10, just wondering if there was anything we'd want to check before moving ahead with this number.

**Review by t-downing — COMMENTED** [review]
Very nice! Just some random little things- main thing is to check that it's taking the right bands from the COGs (this shouldn't be an issue once we're using the ones in PROD though), and whether to use 10 days as the rolling window.

File: exploration/01_baseline_calc_10yr_doy.py (hunk 3 above)
Comment (zackarno) [reply exploration/01_baseline_calc_10yr_doy.py:15]: i had the `%` magic functions, but it was causing issues w/ pre-commit. Have you seen that? I suppose some people have pre-commit ignoring `.ipynb` files, but since this is a `.py` file not sure the right way to deal with it?

File: src/utils/cloud_utils.py (hunk 7 above)
Comment (zackarno) [reply src/utils/cloud_utils.py:72]: ha sorry this is lifted code from [ds-raster-stats](https://github.com/OCHA-DAP/ds-raster-stats). No idea @hannahker  might though? I left it here because it seems like it could be useful

File: exploration/01_baseline_calc_10yr_doy.py (hunk 9 above)
Comment (zackarno) [reply exploration/01_baseline_calc_10yr_doy.py:36]: hmm thats a good point on the `odd` vs `even` number. It's partly a relic of my R-code where the `{terra}` rolling raster stat function `terra::roll(n=10, type = "around")` actually means date +/- 5 days on each side.  However, if that's not happens with `.rolling(time=10)` here, i think we should switch it to have the same effect. The terra roll is a bit peculiar like that as `zoo::roll()` behaves more like you are implying...

7 would also be totally fine, but i think +/-5 days works too

File: requirements.txt
Comment (zackarno) [reply requirements.txt:1]: yeah i had it as somepoint, idk what i did, must have nuked my env too many times -- fixed: [88c91bc](https://github.com/OCHA-DAP/hdx-floodscan/pull/1/commits/88c91bcbbbcbe5963fe26637d6d2e6680777db3f)

File: README.md (hunk 1 above)
Comment (zackarno) [reply README.md]: true! fixed [88c91bc](https://github.com/OCHA-DAP/hdx-floodscan/pull/1/commits/88c91bcbbbcbe5963fe26637d6d2e6680777db3f)

File: src/datasources/floodscan.py (hunk 2 above)
Comment (zackarno) [reply src/datasources/floodscan.py:53]: nice, i will implement something like this... would be nice if there was a more direct solution where you could subset by long_name, i.e  something like:  `ds["sfed"]`, but seems slightly trickier than that

File: exploration/02_wrangle_floodscan.py (hunk 4 above)
Comment (zackarno) [reply exploration/02_wrangle_floodscan.py]: yup good call - fixed: [b238072](https://github.com/OCHA-DAP/hdx-floodscan/pull/1/commits/b238072c7f4c0a677c9c0fba1b3cd029df4e0a96)

File: src/datasources/floodscan.py (hunk 2 above)
Comment (zackarno) [reply src/datasources/floodscan.py:53]: good call -- wrote a little function for this: [b238072](https://github.com/OCHA-DAP/hdx-floodscan/pull/1/commits/b238072c7f4c0a677c9c0fba1b3cd029df4e0a96)

File: src/datasources/floodscan.py (hunk 6 above)
Comment (zackarno) [reply src/datasources/floodscan.py:59]: good to know! i suppose i'll leave it for now if we think its good

File: exploration/01_baseline_calc_10yr_doy.py (hunk 9 above)
Comment (t-downing) [reply exploration/01_baseline_calc_10yr_doy.py:36]: Interesting- the `xr` `rolling` actually takes an asymmetric window when the size is an even number (so with 10, it takes the current day, 4 ahead, and 5 behind). So yeah if we have `center=True`, we should always use an odd number. 

Another option is to not have it be centered. This would mean that we could report the current date's rolling value on that date, instead of waiting to get the whole window (this is how the app is working). 

So basically our options are:
- centered: any odd number of days (e.g. 7, 11)
- not centered (i.e. only looking backward): any number of days

File: src/utils/cloud_utils.py (hunk 7 above)
Comment (hannahker) [reply src/utils/cloud_utils.py:72]: Don't think it's critical! I didn't end up using this function, but think I implemented it this way initially to better handle larger files.

Comment (t-downing) [reply src/utils/cloud_utils.py:72]: Ok good to know! I'll keep that in mind if I have issues with my more simplistic version

File: src/datasources/floodscan.py [hunk 10]
```diff
@@ -52,10 +53,20 @@ def load_floodscan_cogs(
         da_in = process_floodscan_cog(
             cog_name=cog, container_name="global", mode="dev"
         )
-        das.append(da_in)
+        da_sfed = subset_band(da_in, band="SFED")
```
Comment (hannahker) [inline src/datasources/floodscan.py:56]: Nice way to do it!

Comment (t-downing) [reply src/datasources/floodscan.py:56]: Yes I much prefer this to my janky method 😆

File: src/datasources/floodscan.py [hunk 11]
```diff
@@ -52,10 +53,20 @@ def load_floodscan_cogs(
...
+        das.append(da_sfed)
+
     return xr.combine_by_coords(das, combine_attrs="drop")
 
 
+def subset_band(da, band="SFED"):
+    long_name = np.array(da.attrs["long_name"])
+    index_band = np.where(long_name == band)[0] + 1
+    da_subset = da.sel(band=index_band)
```
Comment (t-downing) [inline src/datasources/floodscan.py]: One tiny thing- we may want to use `.isel()` here instead, then we don't rely on the bands being actually called `1` and `2`, and kind of better reflects the `np.where()` method's output (so you don't need to add the ` + 1`).

Comment (zackarno) [reply src/datasources/floodscan.py]: nice, i like it - don't know that i fully understand the difference in the `.sel()` and `.isel()` methods, but it does work an looks nicer! [79d5259](https://github.com/OCHA-DAP/hdx-floodscan/pull/1/commits/79d52590d6856b20c8c18538d4da36867d77c46b)

File: exploration/01_baseline_calc_10yr_doy.py (hunk 9 above)
Comment (zackarno) [reply exploration/01_baseline_calc_10yr_doy.py:36]: yeah that makes sense its also how `zoo::rollmean()` works in R - `terra::(roll, type = "around")` is the odd ball. 

I changed to `11` and kept centered to be consistent with the R work thats already been presented/'endorsed' - [1aa237a](https://github.com/OCHA-DAP/hdx-floodscan/pull/1/commits/1aa237a03e99b08e416fc3c38f9415f45d2db0d4)

File: exploration/01_baseline_calc_10yr_doy.py (hunk 8 above)
Comment (zackarno) [reply exploration/01_baseline_calc_10yr_doy.py:57]: hmm im running in `interactive mode` executing `.py` code line by line in a jupyter kernel.... i guess it might be a bit funky, i wouldn't worry too much about this script as the critical code has already been refactored in `utils/datasources/floodscan.py` and implemented in `exploration/02_wrangle_floodscan.py`... I might even delete this notebook at some point as this work gets further along

**Comment by zackarno** [comment]
So I will go ahead and merge or is anything pending? 

Once merged @hannahker and I can discuss next steps in terms of setting this up as the raster component of HDX pipeline.
